compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs

```diff
@@ -204,6 +227,164 @@ generateDalfRule =
+        let newFinderCache =
+                foldl'
+                    (\fc (im, ifr) -> GHC.extendInstalledModuleEnv fc im ifr) prevFinderCache
+                    $ zip ims ifrs
+        newFinderCacheVar <- liftIO $ newIORef $! newFinderCache
+        modifySession $ \s -> s { hsc_FC = newFinderCacheVar }
+        -- Currently GetDependencies returns things in topological order so A comes before B if A imports B.
+        -- We need to reverse this as GHC gets very unhappy otherwise and complains about broken interfaces.
+        -- Long-term we might just want to change the order returned by GetDependencies
+        mapM_ (uncurry loadDepModule) (reverse deps)
+        (warnings, tcm) <- withWarnings "typecheck" $ \tweak ->
+            GHC.typecheckModule pm { pm_mod_summary = tweak (pm_mod_summary pm) }
+        tcm <- mkTcModuleResult tcm
+        pure (map snd warnings, tcm)
+
+loadDepModule :: GhcMonad m => ModSummary -> ModIface -> m ()
+loadDepModule ms iface = do
+    hsc <- getSession
+    -- The fixIO here is crucial and matches what GHC does. Otherwise GHC will fail
+    -- to find identifiers in the interface and explode.
+    -- For more details, look at hscIncrementalCompile and Note [Knot-tying typecheckIface] in GHC.
+    details <- liftIO $ fixIO $ \details -> do
+        let hsc' = hsc { hsc_HPT = addToHpt (hsc_HPT hsc) (moduleName mod) (HomeModInfo iface details Nothing) }
+        initIfaceLoad hsc' (typecheckIface iface)
+    let mod_info = HomeModInfo iface details Nothing
+    modifySession $ \e ->
+        e { hsc_HPT = addToHpt (hsc_HPT e) (moduleName mod) mod_info }
+    where mod = ms_mod ms
+
+-- TODO Share code with compileModule in ghcide. Given that this is fairly mechanical, this is not critical
+-- but still worth doing in the long-term.
+ondiskDesugar :: HscEnv -> TypecheckedModule -> IO ([FileDiagnostic], Maybe CoreModule)
+ondiskDesugar hsc tm =
+    fmap (either (, Nothing) (second Just)) $
+    runGhcEnv hsc $
+        catchSrcErrors "compile" $ do
+            session <- getSession
+            (warnings, desugar) <- withWarnings "compile" $ \tweak -> do
+                let pm = tm_parsed_module tm
+                let pm' = pm{pm_mod_summary = tweak $ pm_mod_summary pm}
+                let tm' = tm{tm_parsed_module  = pm'}
+                GHC.dm_core_module <$> GHC.desugarModule tm'
+            -- give variables unique OccNames
+            (tidy, details) <- liftIO $ tidyProgram session desugar
+
+            let core = CoreModule
+                         (cg_module tidy)
+                         (md_types details)
+                         (cg_binds tidy)
+                         (mg_safe_haskell desugar)
+
+            return (map snd warnings, core)
+
+-- This rule is for on-disk incremental builds. We cannot use the fine-grained rules that we have for
+-- in-memory builds since we need to be able to serialize intermediate results.
```

Comment:
Why can't more fine-grained rules serialize their respective output?

Reply:
Because GHC does not provide a way to serialize `TypeCheckedModule` or `CoreModule`. It would probably also not help from a performance standpoint to make this too fine-grained. I’ve added a comment.

Reply:
:+1: Thanks for explaining and adding the comment.